C++ Class Support

evelez7 opened the pull request.

evelez7 committed:
[clang][ExtractAPI] Add support for C++ classes

Add ExtractAPI support C++ classes, fields,  methods, and various qualifiers and specifiers

Differential Revision: https://reviews.llvm.org/D153557

daniel-grumberg commented on clang/include/clang/ExtractAPI/API.h:
```diff
@@ -266,6 +274,121 @@ struct StructRecord : APIRecord {
+  virtual void anchor();
+};
+
+struct CXXStaticMethodRecord : CXXMethodRecord {
```
I wonder if these should be more like global functions since this is essentially what they are

evelez7 replied:
Static methods are affected by access modifiers so that's why they inherit from `CXXMethodRecord`.

daniel-grumberg commented on clang/include/clang/ExtractAPI/API.h:
```diff
@@ -444,6 +567,24 @@ struct SymbolReference {
   bool empty() const { return Name.empty() && USR.empty() && Source.empty(); }
 };
 
+struct StaticFieldRecord : APIRecord {
```
This shares a lot of similarities with GlobalVariables

evelez7 replied:
Just changed this to inherit from CXXFieldRecord, static fields also have access.

daniel-grumberg commented on clang/include/clang/ExtractAPI/DeclarationFragments.h:
```diff
@@ -290,7 +297,21 @@ class DeclarationFragmentsBuilder {
   /// base. This template helps reuse the code.
   template <typename FunctionT>
-  static FunctionSignature getFunctionSignature(const FunctionT *);
+  static FunctionSignature getFunctionSignature(const FunctionT *Function) {
```
Is there a reason this needed to move to the header like this

daniel-grumberg commented on clang/include/clang/ExtractAPI/ExtractAPIVisitor.h:
```diff
@@ -280,15 +304,21 @@ bool ExtractAPIVisitorBase<Derived>::VisitEnumDecl(const EnumDecl *Decl) {
-
+bool ExtractAPIVisitorBase<Derived>::WalkUpFromRecordDecl(
+    const RecordDecl *Decl) {
   if (!getDerivedExtractAPIVisitor().shouldDeclBeIncluded(Decl))
```
I don't think this should be in WalkUp* but rather stay in the individual Visit* methods

daniel-grumberg commented on clang/include/clang/ExtractAPI/ExtractAPIVisitor.h:
```diff
@@ -306,22 +336,65 @@ bool ExtractAPIVisitorBase<Derived>::VisitRecordDecl(const RecordDecl *Decl) {
   DeclarationFragments SubHeading =
       DeclarationFragmentsBuilder::getSubHeading(Decl);
-
+  DeclarationFragments Declaration =
```
Is there a reason this moved.

daniel-grumberg commented on clang/include/clang/ExtractAPI/ExtractAPIVisitor.h:
```diff
@@ -306,22 +336,65 @@ bool ExtractAPIVisitorBase<Derived>::VisitRecordDecl(const RecordDecl *Decl) {
+  DeclarationFragments SubHeading =
+      DeclarationFragmentsBuilder::getSubHeading(Decl);
+  APIRecord::RecordKind Kind;
+  if (Decl->isUnion())
```
We should do this in VisitRecordDecl I think

daniel-grumberg commented on clang/include/clang/ExtractAPI/ExtractAPIVisitor.h:
```diff
@@ -306,22 +336,65 @@ bool ExtractAPIVisitorBase<Derived>::VisitRecordDecl(const RecordDecl *Decl) {
+  if (Decl->isUnion())
+    Kind = APIRecord::RecordKind::RK_Union;
+  else if (Decl->isStruct())
+    Kind = APIRecord::RecordKind::RK_Struct;
```
should we have these checks at the beginning and delegate to the relevant methods.

daniel-grumberg commented on clang/lib/ExtractAPI/DeclarationFragments.cpp:
```diff
@@ -492,7 +519,12 @@ DeclarationFragmentsBuilder::getFragmentsForEnum(const EnumDecl *EnumDecl) {
-  return getFragmentsForType(Field->getType(), Field->getASTContext(), After)
+  DeclarationFragments Fragments;
+  if (Field->isMutable())
+    Fragments.append("mutable ", DeclarationFragments::FragmentKind::Keyword);
```
Nice completely forgot this existed!

daniel-grumberg commented on clang/lib/ExtractAPI/DeclarationFragments.cpp:
```diff
@@ -513,6 +545,75 @@ DeclarationFragmentsBuilder::getFragmentsForStruct(const RecordDecl *Record) {
+  if (Record->isStruct())
+    Fragments.append("struct", DeclarationFragments::FragmentKind::Keyword);
+  else if (Record->isUnion())
+    Fragments.append("union", DeclarationFragments::FragmentKind::Keyword);
```
Should this be somewhere else as it isn't C++ specific.

daniel-grumberg commented:
I think this looks pretty good overall. Like we discussed the other day I think we will need to do a follow up patch to clean up how we record things like members to be more flexible with nested classes etc.

aartbik committed:
[mlir][sparse] minor code changes

Submitting for Wren

Reviewed By: K-Wu

Differential Revision: https://reviews.llvm.org/D153804

nicolasvasilache committed:
[mlir][linalg] Add missing op to match the generated file

D141430 added the generated yaml file for (batch_)?matmul_transpose_b ops, but the source of truth core_named_ops.py was not updated.
This change fixes .py file to generate the same result as the yaml file.

Differential revision: https://reviews.llvm.org/D150059

Authored-by: kon72 <kinsei0916@gmail.com>

preames committed:
[RISCV] Regen rvv/fixed-vectors-fmf.ll to avoid spurious test deltas

MaskRay committed:
[dataflow] Fix -Wunused-variable in -DLLVM_ENABLE_ASSERTIONS=off builds after D153006

MaskRay committed:
[bazel] Add includes after 5a63b2b3049e06d83c43c037aab36e9bc3e797cb

nico committed:
[gn] prepare for porting f8927838fa8558702794

evelez7 committed:
Merge branch 'llvm:main' into c++-support
